agent/src/agent/pipeline/json_builder/json_builder.py

```diff
@@ -93,8 +97,12 @@ def edit_multiple(configs: list) -> List[Pipeline]:
 
 def edit(config: dict) -> Pipeline:
     pipeline_ = pipeline.repository.get_by_id(config['pipeline_id'])
-    _load_config(pipeline_, config, is_edit=True)
-    pipeline.manager.update(pipeline_)
+    try:
+        _load_config(pipeline_, config, is_edit=True)
+        pipeline.manager.update(pipeline_)
+    except Exception as e:
+        pipeline.repository.expunge(pipeline_)
+        raise e
```

Comment:
I think writing a context manager in `pipeline.repository` will be a nicer approach. It will expunge in case of exceptions or do save if all is ok. https://book.pythontips.com/en/latest/context_managers.html#implementing-a-context-manager-as-a-class